From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: rmp22 <195054967+rmp22@users.noreply.github.com>
Date: Fri, 1 May 2026 06:31:38 +0800
Subject: [PATCH 162/220] services: Optimizing redundant status bar top hide
 checks

[MikolajQ: Ported to Android 17: the bar is reached through runWithStatusBarIfPresent(), whose result now gates the cached value]

Change-Id: I128348eab541df5b6f0171bfdba402275a5df747
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../server/statusbar/StatusBarManagerService.java  | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/statusbar/StatusBarManagerService.java b/services/core/java/com/android/server/statusbar/StatusBarManagerService.java
index 34aa2d86a..a2fdd33b3 100644
--- a/services/core/java/com/android/server/statusbar/StatusBarManagerService.java
+++ b/services/core/java/com/android/server/statusbar/StatusBarManagerService.java
@@ -213,6 +213,8 @@ public class StatusBarManagerService extends IStatusBarService.Stub implements D
     private int mCurrentUserId;
     private boolean mTracingEnabled;
     private int mLastSystemKey = -1;
+    private volatile boolean mLastTopAppHidesStatusBar;
+    private volatile boolean mLastTopAppHidesStatusBarValid;
 
     private final TileRequestTracker mTileRequestTracker;
 
@@ -234,6 +236,7 @@ public class StatusBarManagerService extends IStatusBarService.Stub implements D
         public void binderDied() {
             mBar.asBinder().unlinkToDeath(this,0);
             mBar = null;
+            mLastTopAppHidesStatusBarValid = false;
             notifyBarAttachChanged();
         }
 
@@ -626,7 +629,14 @@ public class StatusBarManagerService extends IStatusBarService.Stub implements D
                 }
                 return;
             }
-            runWithStatusBarIfPresent(bar -> bar.setTopAppHidesStatusBar(hidesStatusBar));
+            if (mLastTopAppHidesStatusBarValid
+                    && mLastTopAppHidesStatusBar == hidesStatusBar) {
+                return;
+            }
+            if (runWithStatusBarIfPresent(bar -> bar.setTopAppHidesStatusBar(hidesStatusBar))) {
+                mLastTopAppHidesStatusBar = hidesStatusBar;
+                mLastTopAppHidesStatusBarValid = true;
+            }
         }
 
         @Override
@@ -1524,6 +1534,7 @@ public class StatusBarManagerService extends IStatusBarService.Stub implements D
 
         Slog.i(TAG, "registerStatusBar bar=" + bar);
         mBar = bar;
+        mLastTopAppHidesStatusBarValid = false;
         mDeathRecipient.linkToDeath();
         notifyBarAttachChanged();
         final ArrayMap<String, StatusBarIcon> icons;
@@ -1548,6 +1559,7 @@ public class StatusBarManagerService extends IStatusBarService.Stub implements D
 
         Slog.i(TAG, "registerStatusBarForAllDisplays bar=" + bar);
         mBar = bar;
+        mLastTopAppHidesStatusBarValid = false;
         mDeathRecipient.linkToDeath();
         notifyBarAttachChanged();
 
